### PR TITLE
implement duplicate saturation plot code, currently records reads, du…

### PR DESCRIPTION
**Pull request by msettles**
Added a new parameter in super_deduper, log_freq that records the reads/duplicates at a certain interval (default 1M). Then outputs the log in json array of arrays similar to the overlapper histogram.

### CODE REVIEW

File: common/src/counters.h [hunk 1]
```diff
@@ -29,30 +30,19 @@ class Counters {
...
-
-    virtual void input(const SingleEndRead &read) {
-        ++c["TotalFragmentsInput"];
-        ++c["SE_In"]; 
-    }
-
     virtual void input(const ReadBase &read) {
         const PairedEndRead *per = dynamic_cast<const PairedEndRead *>(&read);
         if (per) {
-            input(*per);
```
Comment (joe-angell) [inline common/src/counters.h]: why was this removed?  The functions seem less error prone.

File: common/src/counters.h [hunk 2]
```diff
@@ -441,20 +431,85 @@ class StatsCounters : public Counters {
 };
 
 class SuperDeduperCounters : public Counters {
+
 public:
+    std::vector<std::tuple<int, int>> duplicateProportion;
```
Comment (joe-angell) [inline common/src/counters.h]: recommend we use `uint64_t` for all counters or maybe better would be `uint_fast64_t`

File: common/src/counters.h [hunk 3]
```diff
@@ -441,20 +431,85 @@ class StatsCounters : public Counters {
...
     SuperDeduperCounters() {
         Common();
         c["Duplicate"] = 0;
         c["Ignored"] = 0;
     }
 
+    virtual void input(const ReadBase &read, size_t dup_freq) {
+
+        if (dup_freq > 0 & (c["TotalFragmentsInput"] - c["Ignored"]) % dup_freq == 0){
+            duplicateProportion.push_back(std::make_tuple(c["TotalFragmentsInput"],c["Duplicate"]));
```
Comment (joe-angell) [inline common/src/counters.h]: just curious why you want duplicates over time as apposed to just overall duplicate amount?

File: common/src/counters.h [hunk 4]
```diff
@@ -441,20 +431,85 @@ class StatsCounters : public Counters {
...
+        }
+
+        const PairedEndRead *per = dynamic_cast<const PairedEndRead *>(&read);
```
Comment (joe-angell) [inline common/src/counters.h]: instead of copying this code, you can do Counters::input(read); to call the base class method.

File: common/src/counters.h (hunk 1 above)
Comment (msettles) [reply common/src/counters.h]: seemed simpler, the functions virtual void input(const SingleEndRead &read) and virtual void input(const PairedEndRead &read) were only ever called from within virtual void input(const SingleEndRead &read), even though they are overloaded they never got used outside that one function. So I embedded their functionality in the base function call reducing a function call for each read. I suspect it was thought these functions would get used outside, but they never did and it seems safer not to.

I suppose this comes down to style, but then I'd also have to duplicate the 'duplicate' vector code in both sub-functions as well.

File: common/src/counters.h (hunk 2 above)
Comment (msettles) [reply common/src/counters.h]: K

File: common/src/counters.h (hunk 3 above)
Comment (msettles) [reply common/src/counters.h]: Here is a nice introduction - http://www.cureffi.org/2012/12/11/how-pcr-duplicates-arise-in-next-generation-sequencing/ . Since we are randomly sampling the reads, the first set of reads are expected to have a small proportion of duplicates (most are expected to be original), then as you sequence more you find fewer and fewer original reads to the point where you sequence no new reads and only PCR duplicates so the curve of original reads sequenced start with a steep incline, then levels off and eventually goes flat as you've saturated your sequencing. (similar to https://en.wikipedia.org/wiki/Rarefaction_(ecology) ). Question I want to know is where on the curve am I?

File: common/src/counters.h (hunk 4 above)
Comment (msettles) [reply common/src/counters.h]: thanks de facto c++ newbie again

File: common/src/counters.h (hunk 1 above)
Comment (msettles) [reply common/src/counters.h]: The PE and SE functions effectively ONLY incremented the one counter PE_in or SE_in respectively, so figured you can just as easily do so in within the main input call.

File: common/src/counters.h (hunk 3 above)
Comment (joe-angell) [reply common/src/counters.h]: cool I'll check that out

**Comment by msettles** [comment]
Pull request is ready for final review, merged in Joe's mods, no conflicts, passes make test and my 5M read test dataset.

**Comment by joe-angell** [comment]
lgtm

**Comment by msettles** [comment]
Fixed all confiics with master, ready to merge in with OK @dstreett

**Comment by msettles** [comment]
Seems like I can go ahead and merge this in, I want to work on reformatting the counter and need these changes. Can reopen the issues if needs be.
